src/dataflow/src/source/s3.rs

```diff
@@ -0,0 +1,451 @@
+                .borrow_mut()
+                .push(TimestampMetadataUpdate::StartTimestamping(*id));
+            Some(timestamp_metadata_channel)
+        } else {
+            None
+        }
+    }
+
+    fn get_next_message(
+        &mut self,
+        _consistency_info: &mut ConsistencyInfo,
+        _activator: &Activator,
+    ) -> Result<NextMessage<Out>, anyhow::Error> {
+        if let Some(message) = self.buffer.take() {
+            return Ok(NextMessage::Ready(message));
+        }
+        match self.receiver_stream.try_recv() {
+            Ok(Ok(record)) => {
+                self.offset += 1;
+                Ok(NextMessage::Ready(SourceMessage {
+                    partition: PartitionId::S3,
+                    offset: self.offset.into(),
+                    upstream_time_millis: None,
+                    key: None,
+                    payload: Some(record),
+                }))
+            }
+            Ok(Err(e)) => {
+                log::warn!(
+                    "when reading bucket '{}' for source '{}' ({}): {}",
+                    self.bucket,
+                    self.source_name,
+                    self.id,
+                    e
+                );
+                Err(e)
+            }
+            Err(TryRecvError::Empty) => Ok(NextMessage::Pending),
+            Err(TryRecvError::Disconnected) => Ok(NextMessage::Finished),
+        }
+    }
+
+    fn can_close_timestamp(
+        &self,
+        consistency_info: &ConsistencyInfo,
+        pid: &PartitionId,
+        offset: MzOffset,
+    ) -> bool {
+        if !self.is_activated_reader {
+            true
+        } else {
+            // TODO: when is this ever not true for S3?
+            let last_offset = consistency_info
+                .partition_metadata
+                .get(&pid)
+                // Guaranteed to exist for S3
+                .unwrap()
+                .offset;
+            last_offset >= offset
+        }
+    }
+
+    fn get_worker_partition_count(&self) -> i32 {
+        panic!("s3 sources do not support BYO consistency: get_worker_partition_count")
+    }
+
+    fn has_partition(&self, _partition_id: PartitionId) -> bool {
+        panic!("s3 sources do not support BYO consistency: has_partition")
+    }
+
+    fn ensure_has_partition(&mut self, _consistency_info: &mut ConsistencyInfo, _pid: PartitionId) {
+        panic!("s3 sources do not support BYO consistency: ensure_has_partition")
+    }
+
+    fn update_partition_count(
+        &mut self,
+        consistency_info: &mut ConsistencyInfo,
+        partition_count: i32,
+    ) {
+        log::debug!(
+            "ignoring partition count update type={:?} partition_count={}",
+            consistency_info.source_type,
+            partition_count,
+        )
+    }
+
+    fn buffer_message(&mut self, message: SourceMessage<Out>) {
+        if self.buffer.is_some() {
+            log::error!("Internal error: buffer is not empty, overwriting with new message");
```

Comment:
this should be a panic, IMO. Dropping data and continuing when we have logic errors would be worse than crashing.